### ISSUE TITLE
WandbLogger increases step within epoch

### ISSUE THREAD

**Issue by martinwimpff** [Lightning-AI/lightning#13382]
## 🐛 Bug

If I log train_loss and val_loss with the WandbLogger it increases the step within the epoch s.t. train_loss is logged on steps (1, 3, 5, ..) and val_loss on steps (0, 2, 4, ..).

### To Reproduce
[Colab with BoringModel](https://colab.research.google.com/drive/188qFlb6Q5ujQ15h1JO0xrwd2YfGw1APN?usp=sharing)

cc @awaelchli @morganmcg1 @AyushExel @borisdayma @scottire @manangoel99

**Comment by morganmcg1** [comment]
@manangoel99 can probably help here

**Comment by rohitgr7** [comment]
is this happening on the wandb UI? If yes, can you try updating the x-axis to `trainer/global_step`?

**Comment by manangoel99** [comment]
The reason for this issue is that PL has its own `global_step` and wandb has its own internal `step`.
`global_step` is updated at the end of every training step while the wandb `step` is updated every time `self.log` is called. 

So if you call `self.log` twice within the same training step, it would lead to the wandb `step` incrementing twice even though the `global_step` is incremented only once. 
So as @rohitgr7 mentioned, you would want to change the x-axis to `trainer/global_step` and the plots would fix themselves.

**Comment by martinwimpff** [comment]
Thanks guys!

**Comment by dcharatan** [comment]
It's worth noting that it's not possible to set the x-axis for images. As a result, when using `WandbLogger`, there's no way to reliably show images from a particular training step across runs. For example, if I want to log an output and some visualization, the output will be logged at step `n`, and the visualization will be logged at step `n+1`. There's a set interval at which images are logged (e.g., images are always logged at step `11k`, and visualizations are always logged at step `11k + 1`). If I now add a new logged metric, images are logged at step `12k`, and visualizations are logged at step `12k + 1`. This means that for a training step `k = 77`, the image for the old training run has index 847, and the image for the new training run has index 924, making it nearly impossible to find and compare them.

I've resorted to subclassing `WandbLogger` to override the way the step is specified:

```
@rank_zero_only
def log_metrics(
    self,
    metrics: Mapping[str, float],
    step: int,
) -> None:
    # This is exactly like the regular WandbLogger, except the step must be
    # specified, and images are logged correctly according to the step.
    assert rank_zero_only.rank == 0, "experiment tried to log from global_rank != 0"
    metrics = _add_prefix(metrics, self._prefix, self.LOGGER_JOIN_CHAR)
    self.experiment.log(dict(metrics, **{"trainer/global_step": step}), step=step)
```

Would it be worth making this the default behavior, at least for images?
